### PR TITLE
feat(calibrate): fuzzy title matching in corpus join (#207)

### LINKED ISSUE #207: calibrate: fuzzy title matching in corpus join
## Problem

`quorum calibrate` joins `feedback.jsonl` with `calibrator_traces.jsonl` on `finding_title`, but LLM-generated finding titles are non-deterministic across review runs. Re-reviewing the same file produces different titles for the same underlying issue, so most feedback entries never match a trace.

Current join rate: 329/2,492 (13%). Of 2,000 unmatched entries, the majority have no trace with a matching title — not because the finding wasn't seen, but because the title changed.

Only AST-derived findings (complexity metrics, ast-grep rules) have stable titles. LLM findings like "silent error conversion in load_all" vs ".ok() / .unwrap_or_default() on a fallible I/O result" describe the same issue but never join.

## Proposed Fix

Add fuzzy title matching to `join_feedback_and_traces`:

1. **Exact match first** (current behavior, zero cost)
2. **Token Jaccard fallback** — tokenize titles, compute Jaccard similarity, accept above threshold (e.g. 0.5). Fast, no model dependency.
3. **Optional embedding fallback** — use fastembed (already a dependency) for cosine similarity on unmatched pairs. Higher quality but O(n*m) unless indexed.

Start with token Jaccard — it should recover the majority of matches without adding latency or model dependencies to the calibrate path.

## Impact

With better join rates, `quorum calibrate` produces meaningful thresholds sooner. The current 87% TP / 13% FP class balance may shift significantly when the full corpus participates, potentially making both suppress and boost thresholds computable.

## Evidence

```
Feedback: 2,492 entries (1,355 TP, 464 FP, 425 partial, 248 wontfix)
Traces: 5,962 entries (145 with file_path, 5,817 title-only)
Joined: 329 samples (13% join rate)
Title overlap: 494 of 2,313 unique feedback titles
```

Same file, different run titles:
- Feedback: `.ok() / .unwrap_or_default() on a fallible I/O or parse result silently converts errors`
- Trace: `silent-error-conversion: .ok() / .unwrap_or_default() on a fallible I/O...`

### PR DESCRIPTION
**Pull request by jsnyder**
## Summary

- Add tiered title matching to `join_feedback_and_traces` replacing single-tier exact join
- **Tier 1**: Raw exact `(title, file_path)` — unchanged behavior
- **Tier 2**: Normalized exact — strips backticks, `rule-name:` prefixes, punctuation (except `_`), lowercases, collapses whitespace
- **Tier 3**: Fuzzy same-file — token Jaccard >= 0.5 with best-vs-second margin >= 0.1
- **Tier 4**: Title-only fallback (raw then normalized) with anti-contamination guard preserved
- Add `JoinStats` return type with per-tier counters displayed during `quorum calibrate`
- Rule-prefix stripping requires a hyphen to avoid false positives on common words like `http:`
- Unicode-safe lowercasing (`c.to_lowercase()` not `to_ascii_lowercase()`)

Expected join rate improvement: 13% -> ~25% (329 -> 530+ samples).

## Design Review

- GPT-5.4 reviewed the approach and recommended the tiered structure, normalized-exact tier before Jaccard, and best-vs-second margin framing
- Testing antipatterns expert caught 2 bugs before code was written: Unicode casing asymmetry and rule-prefix false positives

## Test Plan

- [x] 26 new tests (76 total in calibrate module, 1212 total)
- [x] 11 normalize_title tests (backticks, prefixes, underscores, edge cases)
- [x] 7 token_jaccard tests (identity, disjoint, partial, empty, boundary)
- [x] 4 tier-2 tests (backtick recovery, prefix recovery, priority, file-path required)
- [x] 8 tier-3 tests (accept, reject, ambiguity, boundary at 0.5/0.1)
- [x] 3 tier-4 tests (normalized fallback, anti-contamination, short-circuit)
- [x] 2 cross-tier tests (all tiers exercised, accounting identity)
- [x] All existing regression tests updated for new return type
- [x] Clippy clean, release build succeeds

Closes #207

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude Opus 4.6 <noreply@anthropic.com>

## Summary by CodeRabbit

* **New Features**
  * Multi‑tier fuzzy title matching to improve corpus joins: exact, normalized, same‑file fuzzy, and title‑only fallback, with gating to avoid conflicting file‑scoped matches.
  * Join strategy breakdown reporting showing counts per match tier plus skipped/unmatched categories.

* **Documentation**
  * Added design and implementation plans detailing normalization, fuzzy scoring, acceptance criteria, and observability.

* **Tests**
  * Added tests for normalization, Jaccard scoring, thresholds/margins, ambiguity handling, and title‑only gating.

* **Bug Fixes**
  * Improved robustness for malformed/invalid weights and non‑actionable verdicts.

### CODE REVIEW

**Comment by jsnyder** [comment]
@coderabbitai review

File: src/calibrate.rs [hunk 1]
```diff
@@ -9,33 +9,121 @@ use std::collections::{HashMap, HashSet};
...
+/// Minimum margin between best and second-best Jaccard score.
+const FUZZY_AMBIGUITY_MARGIN: f64 = 0.1;
+
+fn normalize_title(raw: &str) -> String {
+    let after_prefix = strip_rule_prefix(raw);
+    let normalized: String = after_prefix
+        .chars()
+        .map(|c| {
+            if c.is_alphanumeric() || c == '_' {
+                c.to_lowercase().next().unwrap_or(c)
```
[the remark this answers is not in the document]
Comment (jsnyder) [reply src/calibrate.rs]: Fixed in 0f4c94e. Switched to `fold + extend(c.to_lowercase())` so multi-codepoint expansions are preserved.

File: src/calibrate.rs [hunk 2]
```diff
@@ -45,72 +133,209 @@ pub fn join_feedback_and_traces(
...
+                                stats.fuzzy_same_file += 1;
+                                continue;
+                            }
+                        }
+                    } else {
+                        stats.ambiguous_skipped += 1;
+                        continue;
+                    }
+                } else if !candidates.is_empty() {
+                    stats.below_threshold += 1;
```
[the remark this answers is not in the document]
Comment (jsnyder) [reply src/calibrate.rs]: Fixed in 0f4c94e. Deferred `below_threshold` accounting with a local flag so it only increments if the title-only fallback also fails.

File: src/calibrate.rs [hunk 3]
```diff
@@ -45,72 +133,209 @@ pub fn join_feedback_and_traces(
...
+                    }
+                    std::collections::hash_map::Entry::Vacant(e) => {
+                        e.insert((tp_w, fp_w));
+                    }
+                }
+            }
+
+            // Tier 3: file -> normalized traces for fuzzy
+            if !norm.is_empty() {
+                file_traces.entry(fp).or_default().push((norm, (tp_w, fp_w)));
```
[the remark this answers is not in the document]
Comment (jsnyder) [reply src/calibrate.rs:189]: Dismissing as false positive. The ambiguous normalized keys in `file_traces` are correctly handled by tier 3's margin check: duplicate normalized titles produce identical Jaccard scores, so `best_score == second_best`, margin is 0 (< 0.1), and `ambiguous_skipped` fires. Removing them from `file_traces` would be wrong -- it would allow matching when we can't distinguish between two traces with the same normalized title for the same file.
